compiler/src/language_server/inlayhint.re

```diff
@@ -33,14 +33,113 @@ let send_no_result = (~id: Protocol.message_id) => {
   Protocol.response(~id, `Null);
 };
 
-let find_hints = program => {
+let build_hint =
+    (position: Protocol.position, message: string): ResponseResult.inlay_hint => {
+  {label: ": " ++ message, position};
+};
+
+let rec resolve_typ = (typ: Types.type_expr) => {
+  switch (typ.desc) {
+  | TTyLink(type_expr)
+  | TTySubst(type_expr) => resolve_typ(type_expr)
+  | _ => typ
+  };
+};
+let string_of_typ = (typ: Types.type_expr) => {
+  Printtyp.string_of_type_scheme(resolve_typ(typ));
+};
+
+let is_func_typ = (typ: Types.type_expr) => {
+  switch (resolve_typ(typ).desc) {
+  | TTyArrow(_, _, _) => true
+  | _ => false
+  };
+};
+
+let find_hints = (~toggle_type_hints: bool, program) => {
   let hints = ref([]);
   open Typedtree;
   open Protocol;
   module Iterator =
     TypedtreeIter.MakeIterator({
       include TypedtreeIter.DefaultIteratorArgument;
-      // Inlay hints for various expressions can be included here.
+
+      let enter_expression = ({exp_desc, exp_type}: expression) =>
+        // Function parameter type hints
+        if (toggle_type_hints) {
+          switch (exp_desc) {
+          | TExpLambda(bindings, _) =>
+            List.iter(
+              ({mb_pat, mb_loc}: match_branch) => {
+                // Get Parameters
+                switch (mb_pat.pat_desc) {
+                | TPatTuple(args) =>
+                  switch (resolve_typ(exp_type).desc) {
+                  | TTyArrow(typ_args, _, _) =>
+                    // Iterate For Types
+                    let argument_typs =
+                      List.map(
+                        ((arg, typ: Types.type_expr)) =>
+                          switch (arg) {
+                          | Default(_) => None
+                          | _ => Some(typ)
+                          },
+                        typ_args,
+                      );
+                    // Make Hints
+                    if (List.length(argument_typs) == List.length(args)) {
+                      List.iter(
+                        ((arg: pattern, typ: option(Types.type_expr))) => {
+                          switch (arg.pat_desc, typ) {
+                          | (TPatVar(_, _), Some(typ)) =>
+                            let bind_end = arg.pat_loc.loc_end;
+                            let p: Protocol.position = {
+                              line: bind_end.pos_lnum - 1,
+                              character: bind_end.pos_cnum - bind_end.pos_bol,
+                            };
+                            let typeSignature = string_of_typ(typ);
+                            hints :=
+                              [build_hint(p, typeSignature), ...hints^];
+                          | _ => ()
+                          }
+                        },
+                        List.combine(args, argument_typs),
+                      );
+                    };
+                  | _ => ()
```

Comment:
I think this would never be possible. Should we do `failwith("Impossible: ...")`?

Reply:
That case can be hit if we are omitting the annotation because `typ` is `None`